### DISCUSSION TITLE
OAuth small issue when directus is included as iframe

### DISCUSSION THREAD

**Discussion by darioguarascio** [directus/directus#8659] [Feature Requests]
Hello, 
I added directus as part of a bigger admin dashboard, including it as an iframe. 
I aslo have a oauth provider enabled, the google one.

When I try to login using oauth, i get a full browser blockage, because of security issues of being contained as iframe. (X-Frame-Options and stuff)
I don't think there is a solution to that, even so, i wouldn't use it. 

It is possible to access using oauth by using a non-iframe browser window, login normally, the, after being logged already, go back to the iframe version, where you would be logged, but it's not the best UX and I can feel how you guys are obsessed with details like this

**Comment by rijkvanzanten** [comment]
> It is possible to access using oauth by using a non-iframe browser window, login normally, the, after being logged already, go back to the iframe version, where you would be logged, but it's not the best UX 

oooooh that's a tricky use case. Even if that redirect would go through correctly, I'm not sure what the situation is nowadays with sharing cookies between frames and the parent page 🤔 

> and I can feel how you guys are obsessed with details like this

Yes, I'm very curious now what our own admin panel looks like where directus is nested within 🤔 Directus was (obviously) designed to be a "full page" so I'm pretty curious how that ends up looking/working :)

**Reply by darioguarascio** [reply]
@rijkvanzanten this is how it looks like on my end
![image](https://user-images.githubusercontent.com/3661363/136704807-53c7dc39-21cc-49ab-bae1-fdd91a0a38fe.png)

and the whole dashboard is already under an oauth layer, which got me thinking if it is possible, given the directus account token, to do a sort of autologin.

**Reply by rijkvanzanten** [reply]
Ahh gotcha. Instead of doing oauth _in_ the iframe, I'm wondering if you might be able to login in this browser session in the current browser context using the regular API endpoints, and then share that session cookie with the iframe somehow 🤔 That way, you can auto-login without any user interaction even

**Reply by darioguarascio** [reply]
@rijkvanzanten I'll give you a practical example.
I'm using cloudflare acces as layer of authentication to load this dashboard. which is hosted on a different url than directus.
I see no way where I could use login api to autologin the user, it would require the user password.
instead, i could load directus as iframe embedding an auth token that I could generate on the fly after authentication and use to generate an iframe url like `https://directus/admin/login?autologin=<token>`

**Reply by rijkvanzanten** [reply]
@darioguarascio You might be able to: 

1. generate a custom token in your access layer
2. pass that token to a custom endpoint in the iframe use (https://directus/custom/login)
3. check if your custom token is what you expected and extract what user goes with it
4. log the user in with that in the custom endpoint
5. redirect to the /admin route (which should now be an authenticated session)

**Reply by darioguarascio** [reply]
@rijkvanzanten yes, that is what I aim to do. What i couldn't find so far is this custom login endpoint. Is it something available already?
I was thinking that it could be redundant when Directus offers a token field for each user, it could be possible to use that one.

**Reply by joselcvarela** [reply]
Hello @darioguarascio 
I think @rijkvanzanten is talking about custom endpoints https://docs.directus.io/guides/api-endpoints/#custom-api-endpoints
You can setup a custom endpoint to do your own logic. In this case a custom authentication.

When registering a custom endpoint, a list of `services` is provided. So you can extract the AuthenticationService and use it to do the login
https://github.com/directus/directus/blob/main/api/src/services/authentication.ts

**Reply by darioguarascio** [reply]
@joselcvarela thanks!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
